Allow OpOverloadPackets as safe torch functions, sanitize dynamo gm before running aotdispatch with cache

jamesjwu opened the pull request:
Summary:
This diff implements two things to improve cache hit rates after testing AOTAutogradCache with internal cogwheel jobs: 
- We should allow torch functions that are OpOverloadPackets
- When running with cache, there are some fields that dynamo puts into the input graph module to aotdispatch that are not stable between runs. We use a context manager to null these out so that they can't be used to affect the output of AOTAutograd, and then we put the fields back onto the gm before returning from AOTAutogradCache.load().

Test Plan:
New unit tests + running nanogpt with AOTAutogradCache. 


Meta:

Run on a long running job
Cache miss: 
 {F1953831996} 

Cache hit: 
 {F1953830872} 

Servicelabs here: 
https://www.internalfb.com/servicelab/experiment/4301352991/


Cache hit: 
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/f660597709-TrainingApplication/attempt_0/version_0/rank_0/index.html


Cache miss: 
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/f660569960-TrainingApplication/attempt_0/version_0/rank_0/index.html

We can see that with these changes, autograd cache hits and saves compile time: 
https://fburl.com/scuba/pt2_compile_events/ycddxstd

Differential Revision: D65436373




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bdhirsh commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -566,107 +595,113 @@ def load(
+                # Special counter when we pass autograd cache but
+                # fail when on inductor guards
+                counters["aot_autograd"]["autograd_cache_guard_miss"] += 1
                 cache_state = "miss"
```
nit: unused?

bdhirsh commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -508,6 +509,34 @@ def wrap_post_compile(
+        saved_fields[field] = getattr(gm, field, None)
+        # Clear the field
+        setattr(gm, field, None)
+    yield
```
nit: this should probably be in a try/finally, in case compilation raises an exception?
```
try:
    yield
finally:
    for field, value in saved_fields.items():
        setattr(gm, field, value)

```

jamesjwu committed:
Allow OpOverloadPackets as safe torch functions, sanitize dynamo gm before running aotdispatch with cache (#139785)

Summary:

This diff implements two things to improve cache hit rates after testing AOTAutogradCache with internal cogwheel jobs: 
- We should allow torch functions that are OpOverloadPackets
- When running with cache, there are some fields that dynamo puts into the input graph module to aotdispatch that are not stable between runs. We use a context manager to null these out so that they can't be used to affect the output of AOTAutograd, and then we put the fields back onto the gm before returning from AOTAutogradCache.load().

Test Plan:
New unit tests + running nanogpt with AOTAutogradCache. 


Meta:

Run on a long running job
Cache miss: 
 {F1953831996} 

Cache hit: 
 {F1953830872} 

Servicelabs here: 
https://www.internalfb.com/servicelab/experiment/4301352991/


Cache hit: 
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/f660597709-TrainingApplication/attempt_0/version_0/rank_0/index.html


Cache miss: 
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/f660569960-TrainingApplication/attempt_0/version_0/rank_0/index.html

We can see that with these changes, autograd cache hits and saves compile time: 
https://fburl.com/scuba/pt2_compile_events/ycddxstd

Reviewed By: bdhirsh

Differential Revision: D65436373
